Design.md

```diff
@@ -238,6 +258,12 @@ This API query S3 credentials of the  user using a `filter` parameter.
 ### List S3 Credentials
 This API list S3 credentials of the  user.
 1. `list-access-keys` api will be called using assumed role credentials.
+2. Retrieve the secret keys, formatted as `<Username>__<AccessKeyID>`, from Redis Sentinel.
+3. Decrypt the secret keys (For more, see [SecretKey-Encryption-Strategy](#SecretKey-Encryption-Strategy)) and add them to the response.
+4. Exclude the keys in the response that are not present in the Redis Sentinel.
```

Comment:
Maybe I will say something stupid, but, can we replace the secret access key with `not available anymore, the access keys can be deleted if unused` and not exclude them. So that vCloud Director Users will be able to clean up what you called `dead keys` directly from vCloud Director UI.

Reply:
According to initial plan when OSE restarts it will invoke `listS3Credentials` and if it receives an empty response, OSE invokes a `createS3Credential` to create a new key for all S3 operations.

With the given idea when OSE restarts and invokes `listS3Credentials`, as Redis restarted, OSIS need to create a new key and include that in the response along with other keys with secret-key as `not available anymore`. Also there is no guarantee OSE will not pick any of the keys with `not available anymore` value. There is no primary key concept in the keys to make sure OSE will always pick given key for S3 operations.

Reply:
I'm not sure to understand, by initial plan you mean our OSIS design right? If yes, why can't we change it? I would prefer that we implement the logic to check if there are no keys in Redis then generate new user keys and insert them in the right place on OSIS/listS3Credentials response so that OSE will pick up the right one to perform the S3 operations. I think this is a lot better than login into the S3 console(which not really possible for a vCloud Director Tenant Users) or using IAM API to cleanup keys.

In my proposition, I'm assuming that the user keys pickup by OSE in OSIS/listS3Credentials response is predictable.

Reply:
Added a new commit for this change 9f4bec93527b6768b4a6d03e6c5dea0932dc4881